LiteCore/Query/N1QL_Parser/n1ql.leg

```diff
@@ -234,11 +234,31 @@ selectExpr =
 
 expr0 =
     x:baseExpr '.' p:propertyPath       { $$ = op("_.", x, p);}
-  | x:baseExpr
-        (_ COLLATE c:collation          { x = collateOp(x, c); }
-            ( _ c2:collation            { extendCollate(x, c2); }
-             )*
-         )?                             { $$ = x; }
+  | x:baseExpr _ co:collateSuffix       { MutableArray coArray = co;
+                                          bool did_collateOp = false;
+                                          for (auto iter = coArray.begin(); iter != coArray.end(); ++iter) {
+                                            if (did_collateOp) {
+                                               extendCollate(x, iter->asstring());
+                                            } else {
+                                               x = collateOp(x, iter->asstring());
+                                               did_collateOp = true;
+                                            }
+                                          }
+                                          $$ = x; }
+  | x:baseExpr                          { $$ = x; }
+
+collateSuffix =
+    co:COLLATE                          { co = Any(); }
+    (
+        c:collation _ !collation        { co = arrayWith(c); }
+      | '(' _
+        ( c:collation _                 { if (co.isNull()) {
+                                            co = arrayWith(c);
+                                          } else {
+                                            appendAny(co, c);
+                                          } }
+        )+ ')' _
+    )                                   { $$ = co; }
```

Comment:
I believe the parser above would work. I'm trying to see if the parser code could process the multiple collations without putting them into an array first. Here is one idea, **but** I haven't tested so it might not work.

```
| x:baseExpr 
    (_ COLLATE                          { bool did_collateOp = false; }
        c:collation _ !collation.       { x = collateOp(x, c); }
            | '(' _
                ( c:collation _         { if (!did_collateOp) {
                                            x = collateOp(x, c);
                                            did_collateOp = true;
                                          } else {
                                            extendCollate(x, c2);
                                          } }
                )+ ')' _
    )?                                  { $$ = x; }
```

Reply:
I see, only the var defined before colon can be seen through the inner blocks.